### PR TITLE
Allow custom versions without VERSIONS definition

### PR DESCRIPTION
**Pull request by fgmacedo**
This PR enables the usage of `django-filebrowser` with [django-image-cropping](https://github.com/jonasundderwolf/django-image-cropping), as proposed in #313.

There's [another PR to the django-image-cropping](https://github.com/jonasundderwolf/django-image-cropping/pull/92) project in order to provide the other side of the integration, with `django-filebrowser` as an backend.

Highlights:

1. Replaces the direct call to `utils.scale_and_crop` by a stack of "processor functions" that are configurable on settings using `FILEBROWSER_PROCESSORS`, with the default pointing to `scale_and_crop`.
- Allow override the `FileBrowseWidget` using `Admin.formfield_overrides` param `widget`.
- `FileBrowseWidget` now renders `data-*` attributes on the `<input>` tag.
- `FileObject.version_generate` param `version_suffix` is not required to be in `VERSIONS` anymore, any suffix should work, if the new `extra_options` param contains enough data to generate a version.
- New concept of `namers` to delegate the construction of a version filename, and deconstruct a version filename back to the original. The `FILEBROWSER_VERSION_NAMER` setting points to the default namer `filebrowser.namers.VersionNamer`.
- `namers.VersionNamer`: A namer that acts like the current behaviour of `django-filebrowser`, generating names based on the `VERSIONS` dict and `version_suffix`.
- `namers.OptionsNamer`: A namer that generates an unique name using the options provided to the `FileObject.version_generate` and the options in `VERSIONS` if an `version_suffix` is provided. Restores the original file name wipping out the last `_version_suffix--plus-any-configs` block entirely.
- Replaces `FileObject.is_version` verfification to only verify if the file is in `VERSIONS_BASEDIR` since with the `namers` there's no guarantee that the version file name is a subset of `VERSIONS` keys.

Motivations behind the design decisions:

1. `FILEBROWSER_PROCESSORS` is required in order to register the [crop_corners](https://github.com/jonasundderwolf/django-image-cropping/blob/master/image_cropping/thumbnail_processors.py#L8) processor. A processor that crops an image based on a box that is stored on [ImageRadioFields](https://github.com/jonasundderwolf/django-image-cropping/blob/master/image_cropping/fields.py#L27).
- Allow override the `FileBrowseWidget` is required because `django-image-cropping` has the custom widget mixin [CropWidget](https://github.com/jonasundderwolf/django-image-cropping/blob/master/image_cropping/widgets.py#L67) that generates meta-data from the source image field (now allowing `FileBrowserField`) and put that meta-data into `data-*` attributes, requiring  the `FileBrowseWidget` to render those attrs into the `<input>` tag.
- Since `django-image-cropping` should potentially generates many versions from the same original source image, there's no concept of `version_prefix`, all information required to generate a version is on the call stack, so I created the `extra_options` to hold that data, that is mixed with the `VERSION` if `version_suffix` exists.
- Any `box` (edges to generate a crop) should generate an unique file name, so the concept of `namers` was used. It's from the [easy_thubnails.namers](https://github.com/SmileyChris/easy-thumbnails/blob/master/easy_thumbnails/namers.py).

Please let me know what do you think, and anything that I could improve.

Best regards,

### CODE REVIEW

**Comment by sehmaschine** [comment]
thanks for the PR.

looks good, but I see 2 potential issues:
a) we add a new setting in-between a version update. not sure that's a good idea.
b) the PR doesn't contain tests.

I could live with a), but b) seems a major issue in my opinion.

**Comment by fgmacedo** [comment]
Thanks for taking time to review It.

a) IMHO, the new settings provide reasonable defaults and the current expected behaviour of `djangp-filebrowser` remains the same.
b) I will update the PR with tests. My first concern was to not break anything, but the changes introduced some new moving parts that need to be covered by the tests.

**Comment by sehmaschine** [comment]
a) agreed.
b) thanks!

**Comment by fgmacedo** [comment]
@sehmaschine Doing some tests, I discovered that `scale_and_crop` returns `False` if both width and height are not provided, or if they are greather than `im` but `upscale` is not present.

    if 'upscale' not in opts:
        if (x < width or not width) and (y < height or not height):
            return False

But on `FileObject._generate_version`, a `Falsy` value of version is converted back to the original `im`:

    version = scale_and_crop(im, ...)
    if not version:
        version = im

Since the `scale_and_crop` is now part of a "process pipeline", do you agree to change Its return value to the received `im` instead of returning `False`?

I added a test that fails on the current `scale_and_crop`:

    # class ScaleAndCropTests(TestCase):
    # ...
    def test_do_not_scale(self):
        version = scale_and_crop(self.im, "", "", "")
        self.assertEqual(version.size, (1000, 750))

This turns `scale_and_crop` optional in the process pipeline.

**Comment by sehmaschine** [comment]
I agree.

**Comment by fgmacedo** [comment]
Hi @sehmaschine. Please let me know if you want more tests and what to test. I think that the integration is already tested with the current suite, so I focused on the new code.

I'm trying to update the docs.

**Comment by fgmacedo** [comment]
@sehmaschine Tests and docs included. Please let me know what do you think.

Best regards,

**Comment by sehmaschine** [comment]
Thanks a lot for the effort. Please be patient with my review – I'll take a look asap.

**Comment by fgmacedo** [comment]
Good news! The related PR on the `django-image-cropping` has been merged! https://github.com/jonasundderwolf/django-image-cropping/pull/92

**Comment by fgmacedo** [comment]
Hi @sehmaschine, how are you? Just pinging to know if there is something that I can do to help you with this.  Best regards!

**Comment by sehmaschine** [comment]
@fgmacedo Sorry for the delay. Lots of client related work right now. I hope being able to update next week.

**Comment by fgmacedo** [comment]
Hi @sehmaschine! How are you? Please let me know if there's anything that I can do for you to help with this. Maybe some other Python/Django related issue to allow you some free time. Best regards!

**Comment by sehmaschine** [comment]
@fgmacedo where exactly are the tests? I don't see any additional tests with the current PR. did I miss something?

**Comment by fgmacedo** [comment]
Hi @sehmaschine !

As pointed out on the comments of this PR on  Apr 15, much of the code changes are refactoring, without adding new behavior, so my work was not to break any test. But I did some changes on the [ScaleAndCropTests](https://github.com/sehmaschine/django-filebrowser/pull/315/files#diff-d81183d781ac117b1198d15ada5fb46eR67) suite.

So I focused in covering the new code:

[FileObjecNamerTests](https://github.com/sehmaschine/django-filebrowser/pull/315/files#diff-d06796e77090d59038f12fe9b1260baeR424)
[ImageProcessorsTests](https://github.com/sehmaschine/django-filebrowser/pull/315/files#diff-d81183d781ac117b1198d15ada5fb46eR33)
[OptionsNamerTests](https://github.com/sehmaschine/django-filebrowser/pull/315/files#diff-d81183d781ac117b1198d15ada5fb46eR33)

**Comment by sehmaschine** [comment]
Some minor questions ...

* Should we use ``VERSION_PROCESSOR`` instead of ``PROCESSOR``?
* With the docs, you say that ``version_generate`` does not require to be in ``VERSIONS``.
But it is required when using ``version_generate``? I don't fully understand this behaviour.

**Comment by sehmaschine** [comment]
Testing looks good so far ... but it seems that ``OptionsNamer`` doesn't work with special characters. If I upload an image ``test_?_image_äää.jpg``, no versions are being generated.

**Comment by fgmacedo** [comment]
> Should we use VERSION_PROCESSOR instead of PROCESSOR

For sure!  I'll update `PROCESSORS` to `VERSION_PROCESSORS` (plural since it's a list).

> With the docs, you say that `version_generate` does not require to be in `VERSIONS`. But it is required when using version_generate? I don't fully understand this behaviour.

The `version_suffix` does not need to be in `VERSIONS`. The name `version_suffix` remains required in order to not change the existing API, I only added an optional parameter. But in fact, if you provide the `extra_options`, the name could be optional. But I like the idea of having a name, even if the name is not required to be in `VERSIONS`. What do you think?


>  If I upload an image `test_?_image_äää.jpg`, no versions are being generated

I will implement a test and fix it.

**Comment by sehmaschine** [comment]
Sorry, but I'm still being confused with ``version_generate``. Your updated docs say that ``version_suffix`` is a required attribute, but it doesn't have to be in ``VERSIONS``. What's the benefit of this change?

Second, you say that any suffix should work if ``extra_options`` contains enough information data to generate a version. Here, you don't say what "enough" means and (IMO) there is a mix up between the suffix being optional for the function versus the suffix being part of ``VERSIONS``.

Maybe this is just a docs issue ... but I somehow still don't get it.

**Comment by sehmaschine** [comment]
All my tests look good so far. Since we now have ``namers``, what do you think about adding a ``TimestampNamer`` which adds a timestamp to the generated version. That way, we could solve possible caching issues when generating versions with an existing filename. I don't wanna add more complexity to this PR – just wanna have your opinion (can be added later anyway).

**Comment by fgmacedo** [comment]
> Sorry, but I'm still being confused with version_generate. Your updated docs say that version_suffix is a required attribute, but it doesn't have to be in VERSIONS. What's the benefit of this change?

I've tryed to explain this on the 3rd "motivation", on the PR description:

> Since django-image-cropping should potentially generates many versions from the same original source image, there's no concept of version_prefix, all information required to generate a version is on the call stack, so I created the extra_options to hold that data, that is mixed with the VERSION if version_suffix exists.

On the `django-filebrowser` side, there's no explicit benefit that I know. But It was required to decouple `version_suffix` from `VERSIONS` in order to allow the usage with `django-image-cropping`.

I think in `version_suffix` as a name that will be used in the version name. And `VERSIONS` is a dict that optionally contains presets of configurations to be applied using `version_suffix` as a key.

We can follow some paths here:

1. I can change the `version_generate` API to turn `version_suffix` optional (a default of `''`).
2. I can change the docs to inform that `version_suffix` is required in order to give versions a base name (even if you are using `OptionsNamer`), and the `VERSIONS` dict holds presets of configurations to be used in `version_generate` if the `VERSIONS` key matches with the `version_suffix` used.

The second point:

> you don't say what "enough" means and (IMO) there is a mix up between the suffix being optional for the function versus the suffix being part of VERSIONS

I agree. I'll update the docs.

An idea: Do not require any value in `extra_options`: If it was empty and there's no key in `VERSIONS` that match `version_suffix`, return the original image as a version. It's the default, if no motifications are required, you'll receive the same image as a version. The process pipeline will be processed, but since there is no options, the result image is the same.

In the future, we could have processors that do simple tasks, like `sepia`, `black and white`, `auto levels`.

**Comment by sehmaschine** [comment]
I'd go with option 2)

Besides, I'd just try to explain the behaviour of ``extra_options`` as simple as possible. We can still extend the docs if people have issues with using the method.

File: filebrowser/base.py [hunk 1]
```diff
@@ -219,13 +219,13 @@ def __init__(self, path, site=None):
         else:
             self.path = path
         self.head = os.path.dirname(path)
-        self.filename = os.path.basename(path)
+        self.filename = force_text(os.path.basename(path))
```
Comment (fgmacedo) [inline filebrowser/base.py]: @sehmaschine  I've found that the problem with `OptionsNamer` was a `''.format(fileobject.filename)`. The `os.path.basename` returns a `str` on python2.7. This fix the problem, but It changes `filename` from `str` to `unicode` in python2.7. 

If you think that this is a bad idea, I can remove it from here and use `force_text` next to the format that I need.

File: docs/fileobject.rst [hunk 2]
```diff
@@ -293,9 +293,13 @@ Version methods
 
 .. method:: version_name(version_suffix, extra_options=None)
 
-    :param version_suffix: A version name as defined in :ref:`settingsversions_versions`.
-    :param extra_options: Optional ``dict`` that will be merged with the
-        predefined ``VERSION`` definition if the given ``version_suffix`` exists.
+    :param version_suffix: A suffix to compose the version name acordinly to
+        the :ref:`settingsversions_version_namer` in use. If the suffix
+        is a key in :ref:`settingsversions_versions` the options will be used.
```
Comment (sehmaschine) [inline docs/fileobject.rst]: a) accordingly
b) what if the suffix is NOT a key in VERSIONS
c) with options you probably mean extra_options, right?

File: docs/fileobject.rst [hunk 3]
```diff
@@ -293,9 +293,13 @@ Version methods
...
+    :param extra_options: An optional ``dict`` to be used in the version
+        generation. If ``version_suffix`` matches a key in
+        :ref:`settingsversions_versions` they will be merged and the
+        ``extra_options`` will take precedence.
```
Comment (sehmaschine) [inline docs/fileobject.rst]: from what I'm seeing with my tests, extra_options do NOT take precedence.

File: filebrowser/base.py (hunk 1 above)
Comment (sehmaschine) [inline filebrowser/base.py]: I'm not sure about this one: With my setup, both self.filename and force_text(self.filename) is unicode (I'm using python 2.7). I'm not exactly sure about the consequences of changing this line ... therefore, I tend to leave it as is. Theoretically speaking, force_text seems to make sense. But I don't wanna run into backwards-incompatible updates here.

**Comment by fgmacedo** [comment]
Hi @sehmaschine, I've updated the docs:

- New section about how to create and use custom processors.
- An attempt to explain in words how the options are computed in `version_generate` (hint: is far more easy to code than to explain!).

Here is the snippet that compute the `options`:

``` python
        options = dict(VERSIONS.get(version_suffix, {}))
        if extra_options:
            options.update(extra_options)
```

So... the keys of `VERSIONS` and `extra_options` are merged into a single dict, and the ones in `extra_options` will take precedence if the same key exists in both dicts.

**Comment by sehmaschine** [comment]
thanks. looks good now ... I'll do a local merge and some cleanup with the docs today.

**Comment by fgmacedo** [comment]
Hi @sehmaschine. Any plans to publish a new release? I'm waiting a new version to update the docs on `django-image-cropping` project.

**Comment by sehmaschine** [comment]
@fgmacedo Well ... the idea was to give other developers the chance for feedback before doing a new release. But since that didn't happen, I'm planning the next release for Tuesday, 9th of August.

**Pull request merged**